### PR TITLE
Fixes 1053 : auto migrations script for ops

### PR DESCRIPTION
**Pull request by lsenta**
- [x] remove dead code in ops
- [x] restructure and clean the ops code
- [x] add automated system to run all migrations since your current db version
- [x] update prepare for testing, etc to manage migrations
- [x] refactor the code to generalize the restore flag => MAINTENANCE mode
  - we'll use this in the rules to disable access to the app during maintenance, cc @ioaNikas 
- [x] block database during migrations / preparation

### CODE REVIEW

**Comment by lsenta** [comment]
@ioaNikas : I removed some of your migration code and I implemented a flag for maintenance we'll use in the rules, please take a look
@Maxstgt : I moved and reorganized some of your code around ops & algolia, please take a look
@Mussuranae : this is the migration tools we've been waiting for! Please take a look, I'll test it tomorrow.

**Comment by lsenta** [comment]
@phpgeekfr I'd like your feedback and I've been told you'll own writing the migrations, please let me know what you think.

**Comment by phpgeekfr** [comment]
@lsenta no problem, Is there something particular that I need to look at ?
If I understand well, now migrations are automated and script know the current DB version by looking in `_META`  ?

**Comment by Maxstgt** [comment]
@lsenta LGTM

File: apps/backend-ops/src/firebaseSetup.ts [hunk 1]
```diff
@@ -114,29 +43,47 @@ export async function restore(appURL: string) {
...
-    console.error(e);
-  }
+  const url = getBackupURL(appURL);
+  const form = {
+    password: process.env['ADMIN_PASSWORD']
+  };
+
+  // promisified request
+  return new Promise((resolve, reject) => {
+    request.post({ url, form }, (error, response) => {
```
Comment (phpgeekfr) [inline apps/backend-ops/src/firebaseSetup.ts]: There is a very good package [request-promise](https://www.npmjs.com/package/request-promise) for this. Maybe overkill for our usage ..

File: apps/backend-ops/src/migrations.ts [hunk 2]
```diff
@@ -1,8 +1,17 @@
...
+import { last } from 'lodash';
+import { IMigrationWithVersion, MIGRATIONS, VERSIONS_NUMBERS } from './firestoreMigrations';
+import { backup } from './firebaseSetup';
+import { appUrl } from '@env';
+import { endMaintenance, startMaintenance } from '../../backend-functions/src/maintenance';
 
 export const VERSION_ZERO = 0;
 
-export interface VersionDoc {
+export interface IVersionDoc {
```
Comment (phpgeekfr) [inline apps/backend-ops/src/migrations.ts]: The Apple touch :)

File: apps/backend-ops/src/migrations.ts [hunk 3]
```diff
@@ -31,3 +37,49 @@ export async function updateDBVersion(db: Firestore, version: number): Promise<a
...
+    version,
+    upgrade: MIGRATIONS[version].update
+  }));
+}
+
+export async function migrate(withBackup: boolean = true) {
+  console.info('start the migration process...');
+  const { db } = loadAdminServices();
+
+  // TODO: disable the database updates
```
Comment (phpgeekfr) [inline apps/backend-ops/src/migrations.ts]: maybe add issue number to your TODOs to retrieve theme easily

File: apps/backend-ops/src/migrations.ts [hunk 4]
```diff
@@ -31,3 +37,49 @@ export async function updateDBVersion(db: Firestore, version: number): Promise<a
...
+    console.info('backup done, moving on to the migrations...');
+  } else {
+    console.warn('⚠️ skipping the backup before running migrations, are you sure?');
+  }
+  try {
+    const currentVersion = await loadDBVersion(db);
+    const migrations = selectAndOrderMigrations(currentVersion);
+    const lastVersion = last(migrations).version;
+
+    console.info(`Running ${migrations.length} between ]${currentVersion}, ${lastVersion}]`);
```
Comment (Mussuranae) [inline apps/backend-ops/src/migrations.ts]: Is there normal to get 2 ] on this line ?
```between ]${currentVersion}, ${lastVersion}]```

**Comment by ioaNikas** [comment]
My first migration code was so old we can say him "ok boomer". Good thing you got rid of it.

File: apps/backend-ops/src/migrations.ts (hunk 3 above)
Comment (lsenta) [reply apps/backend-ops/src/migrations.ts]: Fixed!

File: apps/backend-ops/src/firebaseSetup.ts (hunk 1 above)
Comment (lsenta) [reply apps/backend-ops/src/firebaseSetup.ts]: Agreed, if I use this a third time I'll add it.

File: apps/backend-ops/src/migrations.ts (hunk 2 above)
Comment (lsenta) [reply apps/backend-ops/src/migrations.ts]: :)

File: apps/backend-ops/src/migrations.ts (hunk 4 above)
Comment (lsenta) [reply apps/backend-ops/src/migrations.ts]: Yep, that's the range of migrations, between `]2, 5]` means that it contains `3, 4, 5`

Comment (Mussuranae) [reply apps/backend-ops/src/migrations.ts]: So for this example, the currentVersion 2 will not be included in this range ? 
And if it was [2, 5], the 2nd version would be included in the range ?

File: apps/backend-ops/src/migrations.ts (hunk 2 above)
Comment (lsenta) [reply apps/backend-ops/src/migrations.ts]: (it's a default in create react app configs, it became muscle memory: https://palantir.github.io/tslint/rules/interface-name/)

File: apps/backend-ops/src/migrations.ts (hunk 4 above)
Comment (lsenta) [reply apps/backend-ops/src/migrations.ts]: If the data is already in version 2, you're not going to apply the migration v2.
This is just logging to help you follow along migrations!

**Pull request merged**